### ISSUE TITLE
Add option to populate instance labels with pod names

### ISSUE THREAD

**Issue by SuperQ** [prometheus-operator/prometheus-operator#3652]
**What is missing?**

Friendly `instance` labels.

**Why do we need it?**

In order to improve user-friendliness of the `instance` label, having the pod name as part of the label would be useful.

For example, instead of `instance="10.0.0.1:9090"`, we could have `instance="prometheus-0:9090"`.

**Comment by SuperQ** [comment]
cc @paulfantom

**Comment by lilic** [comment]
Hey, thanks for the feature request. Do you already do this today via relabeling rules in ServiceMonitor and you just want this functionality by default in Prometheus operator?

While the idea for a better UX is a good one, we would have to look at possible downsides to doing this, I am not in favor of adding more and more flags for this as we already have a lot to maintain and causes a lot of confusion to users. So a different approach would be more favorable to me.

cc @prometheus-operator/prometheus-operator-reviewers any more thoughts on this?

**Comment by SuperQ** [comment]
I was looking into doing this with relabeling, but I want to stick with community best practices. If there is a best practice reason to not do this, I probably won't implement it internally.

The one big down-side to this if Kubernetes schedules two pods with the same name in the same namespace. This shouldn't happen in practice, but I don't know if there is anything to specifically prevent this from happening.

**Comment by lilic** [comment]
kube-thanos does already this relabeling https://github.com/thanos-io/kube-thanos/blob/master/jsonnet/kube-thanos/kube-thanos-query.libsonnet#L178-L181

> The one big down-side to this if Kubernetes schedules two pods with the same name in the same namespace. This shouldn't happen in practice, but I don't know if there is anything to specifically prevent this from happening.

Yes in theory that should not happen, it should be prevented as resource names should be unique.

cc @brancz for any backstory on the instance label in prometheus-operator.

**Comment by brancz** [comment]
> The one big down-side to this if Kubernetes schedules two pods with the same name in the same namespace. This shouldn't happen in practice, but I don't know if there is anything to specifically prevent this from happening.

Namespace and pod combination is guaranteed to be unique in Kubernetes (at least on the API level, which is what we're concerned with here), so this is not something we need to worry about, in fact, it could be useful, for StatefulSets for example, as the same pod (like statefulset-0), is intentionally the logical continuation of its potential predecessor, so this would actually be "more correct" than what we have today.

---

There is no real backstory for the value of the `instance` label (I mean .. it's the default), however, the intention of having the `pod` label is quite simple, it makes it trivial to join with kube-state-metrics metrics.

If we do relabel the pod name into the `instance` label, then I think we should do it across the board, and remove the `pod` label that we relabel onto targets today. (There could be a transition period, where the "current"/"old" behavior is still the default, then the new behavior is the default, and finally, the flag would be removed, leaving only the "new" behavior)

What I find particularly compelling about this is that any alerting rule or dashboard can just work with the `instance` label, and don't care about the rest (for example `sum without(instance) ()` type queries), which in some mixins today has to be configurable for this exact reason.

What I don't like about this is that I personally have always been in favor of not enforcing the non-dimensionality of `instance`. This is somewhat the prime example that we need either:

1) Joins on different labels in Prometheus
2) "identifying labels" to be a well-established concept in Prometheus, to allow multiple identifying labels, as well as the non-existance of `instance` (for example being replaced with `pod` in this case).

I honestly don't know what I like better, but without any of them we kind of find ourselves in a dilemma.

My apologies for the long comment.

**Comment by arajkumar** [comment]
IIUC, the request is to replace the ip_addr part from `instance` label and replace it with `pod_name` for better UX. 
Can't we simply keep the `pod` label as is(to help KSM joins) and only replace the `ip_addr` with `pod_name` from `instance` label.? Do you folks feel it would break any of the existing assumptions in alerts and recording rules?

Before:
```
prometheus_engine_queries{container="oauth-proxy", endpoint="web", engine="0", instance="10.128.2.10:9091", job="thanos-querier", namespace="openshift-monitoring", pod="thanos-querier-6d9448674d-qdt8b", service="thanos-querier"}
```

After:
```
prometheus_engine_queries{container="oauth-proxy", endpoint="web", engine="0", instance="thanos-querier-6d9448674d-qdt8b:9091", job="thanos-querier", namespace="openshift-monitoring", pod="thanos-querier-6d9448674d-qdt8b", service="thanos-querier"}
```

IMHO, doing this change would greatly improve the UX for [alerts which only uses job and instance labels](https://github.com/thanos-io/thanos/blob/main/mixin/alerts/sidecar.libsonnet#L16) in the description.

**Comment by paulfantom** [comment]
From my pov IP/port combo as a value of `instance` label is mostly a piece of irrelevant information that we carry. In day-to-day operations, I personally use namespace/pod combo to find out the application metrics and not IP addresses.

In the case of Kubernetes, I don't see much benefit of having `pod:port` as the value of the `instance` label, especially since we already have `endpoint` label. But I am open to trying this.

Personally, I would love to see the `instance` label gone completely and be replaced with `pod`. But that would probably need some work in prometheus.

**Comment by paulfantom** [comment]
I don't see thanos conversation about mixins impacting or being relevant to this issue. Prometheus-operator has a hard dependency on kubernetes so we can simplify things and we don't need to be agnostic.

**Comment by arajkumar** [comment]
I agree that we should migrate to `pod` instead of using `instance`, but as you mentioned it should be done in iterative manner. IMHO keeping the value of `instance` same as `pod` might be a good start.
